Changes proposed by OSI for encoding the featureAttributeTable

RohdeBSH opened the pull request:
The S-102PT has received a proposal from OSI (@pratt-OSI ) on how the attributes of the metadata coverage could be coded in the feature catalogue.

RohdeBSH committed:
Add changes proposed by OSI for encoding the featureAttributeTable #150

RohdeBSH commented:
My XML validation tool did not successfully validate the FeatureCatalogue (FC) with the changes made by OSI.

Could someone else please try to validate the FC? @HG7CS, @hasel001, @anthonyklemm, @pratt-OSI 

### Errors:
- [x] Referenced entity at "https://schemas.s100dev.net/schemas/S100/5.2.0/S100FC/20231214/S100FC.xsd".
- [x] Referenced entity at "https://schemas.s100dev.net/schemas/S100/5.0.0/S100FC/20220610/S100Base.xsd".
- [x] Referenced entity at "https://schemas.s100dev.net/schemas/S100/5.0.0/S100FC/20220610/S100CI.xsd".
- [x] Referenced entity at "https://schemas.s100dev.net/schemas/S100/5.0.0/S100FC/20220610/S100CD.xsd".
- [x] cvc-complex-type.2.4.d: Invalid content was found starting with element “S100FC:definitionReference”. No child element is expected at this point. [122] 
- [x] cvc-complex-type.2.4.a: Invalid content found starting with element “{‘http://www.iho.int/S100FC/5.2’:definitionReference}”. '{‘http://www.iho.int/S100FC/5.2’:uom, "http://www.iho.int/S100FC/5. 2':quantitySpecification, ‘http://www.iho.int/S100FC/5.2’:constraints, ‘http://www.iho.int/S100FC/5.2’:listedValues}' is expected. [132] 
- [x] cvc-complex-type.2.4.a: Invalid content found starting with element “{‘http://www.iho.int/S100FC/5.2’:definitionReference}”: '{‘http://www.iho.int/S100FC/5.2’:uom, "http://www.iho.int/S100FC/5. 2':quantitySpecification, ‘http://www.iho.int/S100FC/5.2’:constraints, ‘http://www.iho.int/S100FC/5.2’:listedValues}' is expected. [142] 
- [x] cvc-complex-type.2.4.a: Invalid content found starting with element “{‘http://www.iho.int/S100FC/5.2’:definitionReference}”.“{‘http://www.iho.int/S100FC/5.2’:listedValues}” is expected. [162]  
- [x] cvc-complex-type.2.4.a: Invalid content found starting with element “{‘http://www.iho.int/S100FC/5.2’:definitionReference}”. “{‘http://www.iho.int/S100FC/5.2’:listedValues}” is expected. [183]
- [x] cvc-complex-type.2.4.a: Invalid content was found starting with element ‚{„http://www.iho.int/S100FC/5.2“:definitionReference}‘ found. '{„http://www.iho.int/S100FC/5.2“:uom, „http://www.iho.int/S100FC/5.2“:quantitySpecification, „http://www.iho.int/S100FC/5.2“: constraints, „http://www.iho.int/S100FC/5.2“:listedValues}' is expected. [193] 
- [x] cvc-complex-type.2.4.a: Invalid content found starting with element ‚{„http://www.iho.int/S100FC/5.2“:definitionReference}‘ found. '{„http://www.iho.int/S100FC/5.2“:uom, „http://www.iho.int/S100FC/5.2“:quantitySpecification, „http://www.iho.int/S100FC/5.2“: constraints, „http://www.iho.int/S100FC/5.2“:listedValues}' is expected. [203] 
- [x] cvc-complex-type.2.4.a: Invalid content found starting with element ‚{„http://www.iho.int/S100FC/5.2“:definitionReference}‘ found. ‚{„http://www.iho.int/S100FC/5.2“:listedValues}‘ is expected. [223] 
- [x] cvc-complex-type.2.4.a: Invalid content found starting with element ‚{„http://www.iho.int/S100FC/5.2“:definitionReference}‘ found. ‚{„http://www.iho.int/S100FC/5.2“:listedValues}‘ is expected. [244] 
- [x] cvc-complex-type.2.4.a: Invalid content found starting with element ‚{„http://www.iho.int/S100FC/5.2“:definitionReference}‘ found. ‚{„http://www.iho.int/S100FC/5.2“:listedValues}‘ is expected. [283] 
- [x] cvc-complex-type.2.4.a: Invalid content found starting with element “{‘http://www.iho.int/S100FC/5.2’:definitionReference}”. “{‘http://www.iho.int/S100FC/5.2’:listedValues}” is expected. [296] 
- [x] cvc-complex-type.2.4.d: Invalid content found starting with element “S100FC:definitionReference”. No child element is expected at this point. [344] 
- [x] cvc-complex-type.2.4.a: Invalid content found starting with element “{‘http://www.iho.int/S100FC/5.2’:subAttributeBinding}”. '{‘http://www.iho.int/S100FC/5.2’:attributeBinding, "http://www.iho.int/S100FC/5. 2':informationBinding, ‘http://www.iho.int/S100FC/5.2’:superType, ‘http://www.iho.int/S100FC/5.2’:subType, ‘http://www.iho.int/S100FC/5.2’:featureUseType}' is expected. [714]
- [x] cvc-identity-constraint.4.3: Key ‘AttributeRef’ with value “null” not found for identity constraint of element ‘S100_FC_FeatureCatalogue’. [809]

RohdeBSH committed:
Add proposed changes by OSI (2025-08-14)

RohdeBSH commented:
FC is now valid...

RohdeBSH commented on sources/3.1.0/FC/102_3_1_0_FC.XML:
```diff
@@ -85,7 +85,565 @@
+				<S100FC:listedValue>
+					<S100FC:label>Unassessed</S100FC:label>
+					<S100FC:definition>Not having been assessed.</S100FC:definition>
+					<S100FC:code>2</S100FC:code>
```
This is incorrect, "Unassessed" has the value "3" according to the IHO registry.

pratt-OSI replied:
There is a disagreement between the S-102 PS and the IHO Registry here.  S-102 PS Table 10-8 states:
<img width="1145" height="143" alt="image" src="https://github.com/user-attachments/assets/5609bcc1-224c-4024-85c3-b0d0f59f1eb8" />
whereas the IHO registry contains conflicting information:
<img width="546" height="90" alt="image" src="https://github.com/user-attachments/assets/d290dd78-966d-45ac-a328-be4f7ec40888" />

which should take precedence?

RohdeBSH replied:
Perfect finding, Jordan.
That's why we do the reviews.
So the question goes to NOAA (@anthonyklemm).
Was it just a small mistake or intentional?
If it was intentional, the IHO registry must be adjusted.
Personally, I would simply correct the transposed digits in S-102.

RohdeBSH replied:
Aligned with the IHO Registry

RohdeBSH commented on sources/3.1.0/FC/102_3_1_0_FC.XML:
```diff
@@ -85,7 +85,565 @@
+				<S100FC:listedValue>
+					<S100FC:label>Assessed (Oceanic)</S100FC:label>
+					<S100FC:definition>The quality of oceanic bathymetric data (depths deeper than 200 metres) has been assessed, however details are not required.</S100FC:definition>
+					<S100FC:code>3</S100FC:code>
```
This is incorrect, "Assessed (Oceanic)" has the value "2" according to the IHO registry.

pratt-OSI replied:
same issue as above

RohdeBSH replied:
Waiting for @anthonyklemm  reply

RohdeBSH replied:
Aligned with the IHO Registry

RohdeBSH commented on sources/3.1.0/FC/102_3_1_0_FC.XML:
```diff
@@ -85,7 +85,565 @@
+				<S100FC:definitionSource ref="IHOREG"/>
+			</S100FC:definitionReference>
+			<S100FC:valueType>real</S100FC:valueType>
+			<S100FC:uom>
```
This attribute has no UOM registered in the IHO registry. But the Format (UOM) was already removed by S-101PT12.

pratt-OSI replied:
I am not an expert on the schema, but it seems harmless to add a UOM for an attribute. However, in this case, it seems essential as the attribute contains a length/size numeric value but give no guidance on the units.  Are the objects 5 cm, 5 m, 5 km, or 5 NM in size?

I recommend the S102PT selected a suitable UOM for this attribute and encodes it in the FC or at least document in the PS.

RohdeBSH replied:
Sounds logical.
But I don't know if that's allowed. I would like to ask Jeff Wootton before we define anything here. In general, I actually believe that IHO registration is mandatory. Some things can be omitted (e.g. as with the Vertical Datum [https://registry.iho.int/fdd/view3.do?idx=996&type=3] ) but nothing can be added.

RohdeBSH replied:
S-102PT25 Minutes:
 JEFF WOOTTON: you
can define these UoM in the Feature Catalogue even if
they are not described in the IHO Data Dictionary
Register, but should be defined at least in the DCEG.

RohdeBSH commented on sources/3.1.0/FC/102_3_1_0_FC.XML:
```diff
@@ -85,7 +85,565 @@
+					<S100Base:upperBound>100.0</S100Base:upperBound>
+					<S100Base:closure>closedInterval</S100Base:closure>
+				</S100CD:range>
+				<S100CD:precision>1</S100CD:precision>
```
This attribute has no precision specified in the IHO registry.

pratt-OSI replied:
This was added as the validation tool was giving a warning to a 'real' value not having precision specified.  This can removed if it is a problem / conflict.  Without it, each ECDIS OEM will display this attribute in their own way.

RohdeBSH replied:
Waiting for Jeff's reply

RohdeBSH replied:
I think this is okay

RohdeBSH commented on sources/3.1.0/FC/102_3_1_0_FC.XML:
```diff
@@ -85,7 +85,565 @@
+				<S100FC:definitionSource ref="IHOREG"/>
+			</S100FC:definitionReference>
+			<S100FC:valueType>real</S100FC:valueType>
+			<S100FC:uom>
```
This attribute has no UOM registered in the IHO registry.

pratt-OSI replied:
The attributes description in the registry states "Percentage of depth that a feature of such size could be detected.", so I added the UOM so the value would be correctly formatted in the pick report, which would be more helpful to the end user.
This can removed if it is a problem / conflict. Without it, each ECDIS OEM will display this attribute in their own way.

RohdeBSH replied:
At least the unit percentage (ID: 383) is registered. Let's see what Jeff has to say about it.

RohdeBSH replied:
S-102PT25 Minutes:
JEFF WOOTTON: you
can define these UoM in the Feature Catalogue even if
they are not described in the IHO Data Dictionary
Register, but should be defined at least in the DCEG.

RohdeBSH commented on sources/3.1.0/FC/102_3_1_0_FC.XML:
```diff
@@ -85,7 +85,565 @@
+				<S100FC:definitionSource ref="IHOREG"/>
+			</S100FC:definitionReference>
+			<S100FC:valueType>real</S100FC:valueType>
+			<S100FC:uom>
```
This attribute has no UOM registered in the IHO registry.

pratt-OSI replied:
The IHO registry gives the 'Quantity Specification' as 'length', which implies that the attribute value will need a unit.  Otherwise, the attribute contents are ambiguous.
I recommend the S102PT selected a suitable UOM for this attribute and encodes it in the FC or at least document in the PS.

RohdeBSH replied:
Waiting for Jeff's reply

RohdeBSH replied:
S-102PT25 Minutes:
JEFF WOOTTON: you
can define these UoM in the Feature Catalogue even if
they are not described in the IHO Data Dictionary
Register, but should be defined at least in the DCEG.

RohdeBSH commented on sources/3.1.0/FC/102_3_1_0_FC.XML:
```diff
@@ -85,7 +85,565 @@
+					<S100Base:lowerBound>0</S100Base:lowerBound>
+					<S100Base:closure>gtSemiInterval</S100Base:closure>
+				</S100CD:range>
+				<S100CD:precision>2</S100CD:precision>
```
precision is "1" according to the IHO registry

pratt-OSI replied:
Noted, I will update this in my proposed S102_FC.xml

RohdeBSH replied:
Corrected in commit 002ecbc51e9d622775ad572f62b84e8f0a91b70a

RohdeBSH commented on sources/3.1.0/FC/102_3_1_0_FC.XML:
```diff
@@ -85,7 +85,565 @@
+			</S100FC:definitionReference>
+			<S100FC:valueType>text</S100FC:valueType>
+			<S100FC:constraints>
+				<S100CD:stringLength>256</S100CD:stringLength>
```
That is not defined in the IHO registry

pratt-OSI replied:
Yes, I added the stringLength constraint to make it easier on implementers.  This can be changed or removed if it is an issue.

RohdeBSH replied:
Waiting for Jeff's reply

RohdeBSH commented on sources/3.1.0/FC/102_3_1_0_FC.XML:
```diff
@@ -85,7 +85,565 @@
+			</S100FC:definitionReference>
+			<S100FC:valueType>text</S100FC:valueType>
+			<S100FC:constraints>
+				<S100CD:stringLength>256</S100CD:stringLength>
```
The length is specified with "100" not "256" in the IHO registry

pratt-OSI replied:
Noted, I will update this in my proposed S102_FC.xml

RohdeBSH replied:
Corrected in commit 9a7f334967b2233f2cd415b8e93d51a1b1984378

RohdeBSH commented on sources/3.1.0/FC/102_3_1_0_FC.XML:
```diff
@@ -127,6 +685,135 @@
+			<S100FC:attributeBinding sequential="false">
+				<S100FC:multiplicity>
+					<S100Base:lower>0</S100Base:lower>
+					<S100Base:upper xsi:nil="false" infinite="false">1</S100Base:upper>
```
I don't know why you need that feature, but upper bound "1" is not correct. With the multiple vertical datum structure this can be overwritten at every BathymetryCoverage.

pratt-OSI replied:
I will demonstrate my proposed implementation at the upcoming S102PT meeting.  I am generating one VerticalDatumOfData feature object for each BathymetryCoverage.nn.  This allows each coverage to have its own VerticalDatumOfData feature with a single verticalDatum attribute.  Most importantly, this would be available in the ECDIS pick report.

RohdeBSH commented on sources/3.1.0/FC/102_3_1_0_FC.XML:
```diff
@@ -127,6 +685,135 @@
+			<S100FC:permittedPrimitives>coverage</S100FC:permittedPrimitives>
+		</S100FC:S100_FC_FeatureType>
+		<S100FC:S100_FC_FeatureType isAbstract="false">
+			<S100FC:name>Vertical Datum of Data</S100FC:name>
```
Where is this defined in the S-102 spec?

pratt-OSI replied:
I liberated the VerticalDatumOfData feature object from the S-101 FC.

I needed a feature object to store the vertical datum attribute and I could not use QualityOfBathymetryCoverage because the spatial extents of the BathymetryCoverage.nn vs featureAttributeTable survey areas are not the same.

I will demonstrate my proposed implementation at the upcoming S102PT meeting.

RohdeBSH reviewed (commented):
The content needs a review, some specifications are incorrect.

RohdeBSH committed:
Constraint corrected -> Length 100

RohdeBSH committed:
Corrected precision -> 1

RohdeBSH committed:
Change attribute name 'bathyCoverage' to 'bathyCoverageAchieved'

RohdeBSH committed:
FC changes by OSI; add complex attributes; 20250821

RohdeBSH committed:
corrected the values of dataassessment

RohdeBSH committed:
As requested by S-102PT25 change attribute 'typeOfBathymetricEstimationUncertainty' to 'private' #129

RohdeBSH committed:
Reorganize structure -> print pretty

RohdeBSH committed:
Update dataAssessment status for oceanic features

There was an error in S-102 regarding the ‘dataAssessment’ attribute. The CodeList values were not compliant with the IHO Registry. This has now been corrected, and the FC and PS text now match.
